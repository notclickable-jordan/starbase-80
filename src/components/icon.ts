import { is } from "../shared/is";
import { IconAspect } from "../shared/types";
import { Anchor } from "./anchor";

const iconColors = [
	"blue",
	"rose",
	"green",
	"red",
	"yellow",
	"cyan",
	"pink",
	"orange",
	"sky",
	"slate",
	"emerald",
	"zinc",
	"neutral",
	"amber",
	"violet",
];
const iconLevel = 300;

const getIconColor = (index: number) => `bg-${iconColors[iconColors.length % index]}-${iconLevel}`;

interface IProps {
	name: string;
	index: number;
	icon?: string;
	iconColor?: string;
	iconBG?: string;
	iconBubble?: boolean;
	iconBubblePadding?: boolean;
	iconAspect?: IconAspect;
	uri?: string;
	newWindow?: boolean;
	categoryBubblePadding?: boolean;
}

export const Icon = function (props: IProps): string {
	const {
		name,
		uri,
		icon,
		index,
		iconBG,
		iconBubble,
		iconBubblePadding,
		iconColor,
		iconAspect,
		newWindow,
		categoryBubblePadding,
	} = props;

	if (is.null(icon)) {
		if (!is.null(uri)) {
			return Anchor({ uri: uri as string, title: name, newWindow, children: IconBlank({ index }) });
		}

		return IconBlank({ index });
	}

	let bubblePadding = categoryBubblePadding || false;

	if (iconBubblePadding === true) {
		bubblePadding = true;
	} else if (iconBubblePadding === false) {
		bubblePadding = false;
	}

	return IconBase({
		icon: icon as string,
		iconBG,
		iconColor,
		iconBubble,
		iconBubblePadding: bubblePadding,
		iconAspect,
	});
};

interface IIconBlankProps {
	index: number;
}

function IconBlank(props: IIconBlankProps) {
	const { index } = props;
	return `
		<span
			class="${`block w-16 h-16 rounded-2xl border border-black/5 shadow-sm ${getIconColor(index)} overflow-hidden`}"
		/>
	`;
}

enum IconType {
	uri,
	material,
	dashboard,
	selfhst,
}

interface IIconBaseProps {
	icon: string;
	iconColor?: string;
	iconBG?: string;
	iconBubble?: boolean;
	iconBubblePadding?: boolean;
	iconAspect?: IconAspect;
}

function IconBase(props: IIconBaseProps) {
	let { icon, iconBG, iconBubble, iconBubblePadding, iconColor, iconAspect = "square" } = props;

	let iconType: IconType = IconType.uri;

	if (icon.startsWith("http") || icon.startsWith("/")) {
		iconType = IconType.uri;
	} else if (icon.startsWith("mdi-")) {
		iconType = IconType.material;
	} else if (icon.startsWith("selfhst-")) {
		iconType = IconType.selfhst;
	} else {
		iconType = IconType.dashboard;
	}

	// Everyone starts the same size
	let bubbleClassName = "flex items-center justify-center overflow-hidden bg-contain";
	let iconWrapperWidthHeightClassName = "";
	let iconItselfWidthHeightClassName = "";

	switch (iconAspect) {
		case "width":
			iconItselfWidthHeightClassName = "w-16";
			iconWrapperWidthHeightClassName += " w-16";

			if (iconBubblePadding) {
				iconItselfWidthHeightClassName = "w-14";
			}
			break;
		case "height":
			iconItselfWidthHeightClassName = "h-16";
			iconWrapperWidthHeightClassName += " h-16";

			if (iconBubblePadding) {
				iconItselfWidthHeightClassName = "h-14";
			}
			break;
		case "square":
		default:
			iconItselfWidthHeightClassName = "w-16 h-16 object-contain";
			iconWrapperWidthHeightClassName += " w-16 h-16";

			if (iconBubblePadding) {
				iconItselfWidthHeightClassName = "w-14 h-14 object-contain";
			}
			break;
	}

	bubbleClassName += iconWrapperWidthHeightClassName;

	if (is.null(iconBubble) || iconBubble !== false) {
		bubbleClassName += " rounded-2xl border border-black/5 shadow-sm";
	}

	const bubbleStyle: string[] = [];

	switch (iconType) {
		case IconType.uri:
		case IconType.dashboard:
		case IconType.selfhst:
			// Default to bubble and no background for URI, Dashboard and selfhst icons
			if (!is.null(iconBG)) {
				if (iconBG?.startsWith("#")) {
					bubbleStyle.push(`background-color: ${iconBG}`);
				} else {
					bubbleClassName += ` bg-${iconBG}`;
				}
			}
			break;
		case IconType.material:
			// Material icons get a color and a background by default, then an optional bubble
			if (is.null(iconBG)) {
				bubbleClassName += ` bg-slate-200 dark:bg-gray-900`;
			} else {
				if (iconBG?.startsWith("#")) {
					bubbleStyle.push(`background-color: ${iconBG}`);
				} else {
					bubbleClassName += ` bg-${iconBG}`;
				}
			}

			if (is.null(iconColor)) {
				iconColor = "black dark:bg-white";
			}
			break;
	}

	const mdiIconStyle: string[] = [];
	let mdiIconColorFull = "bg-" + iconColor;

	if (!is.null(iconColor) && iconColor?.startsWith("#")) {
		mdiIconColorFull = "";
		mdiIconStyle.push(`background-color: ${iconColor}`);
	}

	switch (iconType) {
		case IconType.uri:
			return `<span class="${bubbleClassName}" style="${unwrapStyles(
				bubbleStyle
			)}"><img src="${icon}" alt="" class="${iconItselfWidthHeightClassName || ""}" /></span>`;
		case IconType.dashboard:
			icon = icon.replace(".png", "").replace(".jpg", "").replace(".svg", "");
			return `<span class="${bubbleClassName}" style="${unwrapStyles(bubbleStyle)}">
				<img
					src=${`https://cdn.jsdelivr.net/gh/walkxcode/dashboard-icons/png/${icon}.png`}
					alt=""
					class="${iconItselfWidthHeightClassName || ""}"
				/>
			</span>`;
		case IconType.selfhst:
			icon = icon.replace("selfhst-", "").replace(".png", "").replace(".svg", "");
			return `<span class="${bubbleClassName}" style="${unwrapStyles(bubbleStyle)}">
				<img
					src=${`https://cdn.jsdelivr.net/gh/selfhst/icons/png/${icon}.png`}
					alt=""
					class="${iconItselfWidthHeightClassName || ""}"
				/>
			</span>`;
		case IconType.material:
			icon = icon.replace("mdi-", "").replace(".svg", "");
			return `
			<span class="${bubbleClassName}" style="${unwrapStyles(bubbleStyle)}">
				<span>
					<span
						class="flex items-center justify-center ${iconItselfWidthHeightClassName} ${mdiIconColorFull} overflow-hidden"
						style="${unwrapStyles(
							mdiIconStyle.concat([
								`mask: url(https://cdn.jsdelivr.net/npm/@mdi/svg@latest/svg/${icon}.svg) no-repeat center / contain`,
								`webkit-mask: url(https://cdn.jsdelivr.net/npm/@mdi/svg@latest/svg/${icon}.svg) no-repeat center / contain`,
							])
						)}"
					></span>
				</span>
				</span>
			`;
	}
}

function unwrapStyles(styles: string[]): string {
	if (is.null(styles)) {
		return "";
	}

	return styles.join(";");
}
