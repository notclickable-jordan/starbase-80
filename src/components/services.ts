import { is } from "../shared/is";
import { IService } from "../shared/types";
import { Anchor } from "./anchor";
import { Icon } from "./icon";

interface IServicesProps {
	services: IService[];
	categoryBubblePadding?: boolean;
}

export const Services = function (props: IServicesProps) {
	const { services, categoryBubblePadding } = props;

	return `
		<ul class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-1 lg:gap-2 lg:gap-y-4">
			${services.map((service, index) => Service({ index, service, categoryBubblePadding })).join("")}
		</ul>
	`;
};

interface IServiceProps {
	service: IService;
	index: number;
	categoryBubblePadding?: boolean;
}

function Service(props: IServiceProps) {
	const { service, index, categoryBubblePadding } = props;

	const {
		name,
		uri,
		description,
		icon,
		iconLight,
		iconDark,
		iconBG,
		iconBubble,
		iconBubblePadding,
		iconColor,
		iconAspect,
		newWindow,
	} = service;

	return `
		<li class="p-4">
		${Anchor({
			uri,
			newWindow,
			className: "flex gap-4",
			children: `${
				!is.null(icon)
					? `<span class="flex-shrink-0 flex">${Icon({
							name,
							icon,
							iconLight,
							iconDark,
							uri,
							index,
							iconColor,
							iconBG,
							iconBubble,
							iconAspect,
							newWindow,
							categoryBubblePadding,
							iconBubblePadding,
						})}</span>`
					: ``
			}
		<div>
			<h3 class="text-lg mt-1 font-semibold line-clamp-1">${name}</h3>
			${!is.null(description) ? `<p class="text-sm text-black/50 dark:text-white/50 line-clamp-1">${description}</p>` : ``}
		</div>`,
		})}
			
		</li>
	`;
}
