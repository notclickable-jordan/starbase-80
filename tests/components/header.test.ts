import { Header } from "../../src/components/header";

describe("Header Component", () => {
	const testColors = ["blue", "rose", "green", "red", "yellow"];
	const longString =
		"This is a very long header title that tests how the component handles extensive text content and ensures that it properly renders without breaking the layout or functionality of the header element";

	describe("Required props only", () => {
		test("should render with no props", () => {
			const result = Header({});
			expect(result).toMatchSnapshot();
		});
	});

	describe("With title variations", () => {
		test("should render with short title", () => {
			const result = Header({
				title: "My App",
			});
			expect(result).toMatchSnapshot();
		});

		test("should render with long title", () => {
			const result = Header({
				title: longString,
			});
			expect(result).toMatchSnapshot();
		});

		testColors.forEach(color => {
			test(`should render with ${color} themed title`, () => {
				const result = Header({
					title: `${color.charAt(0).toUpperCase() + color.slice(1)} Dashboard`,
				});
				expect(result).toMatchSnapshot();
			});
		});

		test("should render with HTML-like title", () => {
			const result = Header({
				title: "App <em>with</em> emphasis & special chars: @#$%^&*()",
			});
			expect(result).toMatchSnapshot();
		});
	});

	describe("With icon variations", () => {
		test("should render with simple icon path", () => {
			const result = Header({
				icon: "/logo.png",
			});
			expect(result).toMatchSnapshot();
		});

		test("should render with full URL icon", () => {
			const result = Header({
				icon: "https://example.com/icon.png",
			});
			expect(result).toMatchSnapshot();
		});

		test("should render with long icon path", () => {
			const result = Header({
				icon: "/very/long/path/to/assets/images/icons/application/header/logo-with-very-long-filename.png",
			});
			expect(result).toMatchSnapshot();
		});

		testColors.forEach(color => {
			test(`should render with ${color} themed icon`, () => {
				const result = Header({
					icon: `/icons/${color}-theme-logo.svg`,
				});
				expect(result).toMatchSnapshot();
			});
		});
	});

	describe("Combined prop variations", () => {
		test("should render with both title and icon", () => {
			const result = Header({
				title: "My Application",
				icon: "/logo.png",
			});
			expect(result).toMatchSnapshot();
		});

		testColors.forEach((color, index) => {
			test(`should render with ${color} theme combination`, () => {
				const result = Header({
					title: `${color.charAt(0).toUpperCase() + color.slice(1)} Portal`,
					icon: `/icons/${color}-logo.png`,
				});
				expect(result).toMatchSnapshot();
			});
		});

		test("should render with long strings for both props", () => {
			const result = Header({
				title: longString,
				icon: "/very/long/path/to/assets/images/icons/application/header/main-application-logo-with-very-descriptive-filename.png",
			});
			expect(result).toMatchSnapshot();
		});

		test("should render with special characters in title", () => {
			const result = Header({
				title: 'Appâ„¢ & Co. - "The Best" <Solution> for [Advanced] Users!',
				icon: "/special-logo.svg",
			});
			expect(result).toMatchSnapshot();
		});

		test("should render with empty strings", () => {
			const result = Header({
				title: "",
				icon: "",
			});
			expect(result).toMatchSnapshot();
		});
	});
});
