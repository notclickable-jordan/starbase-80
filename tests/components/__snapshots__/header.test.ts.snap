// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Header Component Combined prop variations should render with blue theme combination 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/icons/blue-logo.png" alt="Blue Portal" class="inline-block w-16 h-16" />
			<h1>Blue Portal</h1>
		</div>
	"
`;

exports[`Header Component Combined prop variations should render with both title and icon 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/logo.png" alt="My Application" class="inline-block w-16 h-16" />
			<h1>My Application</h1>
		</div>
	"
`;

exports[`Header Component Combined prop variations should render with empty strings 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			
			<h1></h1>
		</div>
	"
`;

exports[`Header Component Combined prop variations should render with green theme combination 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/icons/green-logo.png" alt="Green Portal" class="inline-block w-16 h-16" />
			<h1>Green Portal</h1>
		</div>
	"
`;

exports[`Header Component Combined prop variations should render with long strings for both props 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/very/long/path/to/assets/images/icons/application/header/main-application-logo-with-very-descriptive-filename.png" alt="This is a very long header title that tests how the component handles extensive text content and ensures that it properly renders without breaking the layout or functionality of the header element" class="inline-block w-16 h-16" />
			<h1>This is a very long header title that tests how the component handles extensive text content and ensures that it properly renders without breaking the layout or functionality of the header element</h1>
		</div>
	"
`;

exports[`Header Component Combined prop variations should render with red theme combination 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/icons/red-logo.png" alt="Red Portal" class="inline-block w-16 h-16" />
			<h1>Red Portal</h1>
		</div>
	"
`;

exports[`Header Component Combined prop variations should render with rose theme combination 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/icons/rose-logo.png" alt="Rose Portal" class="inline-block w-16 h-16" />
			<h1>Rose Portal</h1>
		</div>
	"
`;

exports[`Header Component Combined prop variations should render with special characters in title 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/special-logo.svg" alt="App™ & Co. - "The Best" <Solution> for [Advanced] Users!" class="inline-block w-16 h-16" />
			<h1>App™ & Co. - "The Best" <Solution> for [Advanced] Users!</h1>
		</div>
	"
`;

exports[`Header Component Combined prop variations should render with yellow theme combination 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/icons/yellow-logo.png" alt="Yellow Portal" class="inline-block w-16 h-16" />
			<h1>Yellow Portal</h1>
		</div>
	"
`;

exports[`Header Component Required props only should render with no props 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			
			<h1>undefined</h1>
		</div>
	"
`;

exports[`Header Component With icon variations should render with blue themed icon 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/icons/blue-theme-logo.svg" alt="" class="inline-block w-16 h-16" />
			<h1>undefined</h1>
		</div>
	"
`;

exports[`Header Component With icon variations should render with full URL icon 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="https://example.com/icon.png" alt="" class="inline-block w-16 h-16" />
			<h1>undefined</h1>
		</div>
	"
`;

exports[`Header Component With icon variations should render with green themed icon 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/icons/green-theme-logo.svg" alt="" class="inline-block w-16 h-16" />
			<h1>undefined</h1>
		</div>
	"
`;

exports[`Header Component With icon variations should render with long icon path 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/very/long/path/to/assets/images/icons/application/header/logo-with-very-long-filename.png" alt="" class="inline-block w-16 h-16" />
			<h1>undefined</h1>
		</div>
	"
`;

exports[`Header Component With icon variations should render with red themed icon 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/icons/red-theme-logo.svg" alt="" class="inline-block w-16 h-16" />
			<h1>undefined</h1>
		</div>
	"
`;

exports[`Header Component With icon variations should render with rose themed icon 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/icons/rose-theme-logo.svg" alt="" class="inline-block w-16 h-16" />
			<h1>undefined</h1>
		</div>
	"
`;

exports[`Header Component With icon variations should render with simple icon path 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/logo.png" alt="" class="inline-block w-16 h-16" />
			<h1>undefined</h1>
		</div>
	"
`;

exports[`Header Component With icon variations should render with yellow themed icon 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			<img src="/icons/yellow-theme-logo.svg" alt="" class="inline-block w-16 h-16" />
			<h1>undefined</h1>
		</div>
	"
`;

exports[`Header Component With title variations should render with HTML-like title 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			
			<h1>App <em>with</em> emphasis & special chars: @#$%^&*()</h1>
		</div>
	"
`;

exports[`Header Component With title variations should render with blue themed title 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			
			<h1>Blue Dashboard</h1>
		</div>
	"
`;

exports[`Header Component With title variations should render with green themed title 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			
			<h1>Green Dashboard</h1>
		</div>
	"
`;

exports[`Header Component With title variations should render with long title 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			
			<h1>This is a very long header title that tests how the component handles extensive text content and ensures that it properly renders without breaking the layout or functionality of the header element</h1>
		</div>
	"
`;

exports[`Header Component With title variations should render with red themed title 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			
			<h1>Red Dashboard</h1>
		</div>
	"
`;

exports[`Header Component With title variations should render with rose themed title 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			
			<h1>Rose Dashboard</h1>
		</div>
	"
`;

exports[`Header Component With title variations should render with short title 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			
			<h1>My App</h1>
		</div>
	"
`;

exports[`Header Component With title variations should render with yellow themed title 1`] = `
"
		<div class="p-2 xl:p-4 flex flex-nowrap justify-center items-center gap-2 xl:flex-wrap">
			
			<h1>Yellow Dashboard</h1>
		</div>
	"
`;
